diff --git a/configure b/configure
index 2ca6316..176b931 100755
--- a/configure
+++ b/configure
@@ -3880,8 +3880,8 @@ fi
 case "${BUILD_32BIT}" in
   yes|default)
 	echo "Building 32-bit pk"
-	CFLAGS="$default_CFLAGS -m32"
-	LDFLAGS="-m32"
+	CFLAGS="$default_CFLAGS"
+	LDFLAGS=
 	install_subdir="riscv32-unknown-elf"
 	;;
   *)
diff --git a/configure.ac b/configure.ac
index 1bf0326..2ae7d43 100644
--- a/configure.ac
+++ b/configure.ac
@@ -86,8 +86,8 @@ AC_ARG_ENABLE([32bit],
 case "${BUILD_32BIT}" in
   yes|default)
 	echo "Building 32-bit pk"
-	CFLAGS="$default_CFLAGS -m32"
-	LDFLAGS="-m32"
+	CFLAGS="$default_CFLAGS"
+	LDFLAGS=
 	install_subdir="riscv32-unknown-elf"
 	;;
   *)
